### DISCUSSION TITLE
How can I extract the start, stop and step values from a py::slice object?

### DISCUSSION THREAD

**Discussion by 123epsilon** [pybind/pybind11#4716] [Q&A] [answered]
I have a custom class with a `__getitem__` implementation that requires custom behavior for slices. I.e I am overloading the following: `__getitem__(py::slice sliceObj)`, how can I retrieve the start, stop, and step attributes from the py::slice object? I'd like to have them as either a py::none instance or as an int. Using the `getattr()` function returns me an object, but I don't know how to parse that into the type I want.

**Comment by 123epsilon** [comment, marked as answer]
Solution:

```
auto parseAttr = [](const py::object &obj) -> OptionalIntValue {
  if(py::isinstance<py::none>(obj)){
    return py::none();
  }
  else if(py::isinstance<py::int_>(obj)){
    return obj.cast<int>();
  }
  else {
    throw std::invalid_argument("Invalid indices);
  }
  };
        
// py::isinstance<py::slice>(sliceObj) == true
parseAttr(getattr(sliceObj, "start"));
```

**Reply by reubenjohn** [reply]
Can you explain your answer a bit more?

I am trying to overload the __getitem__ method of MyList. Do you know how to implement it?

```python
py::class_<MyList>(m, "MyList")
          .def("__getitem__", [](const MyList& s, const py::slice& slice) {
            MyList sliced;
            // if ( slice[1] == py::None ) {}
            // sliced.insert(sliced.end(), s.begin() + slice[0], s.begin() + slice[1]);
            return sliced;
        })

```

**Reply by 123epsilon** [reply]
My answer above is just how one would parse each attribute in the slice object to determine whether it is a None or an integer. Then you'd implement your custom behavior using that function. So for instance in the above example, you might use `parseAttr` as follows to get the start, stop, and step of the slice object your user passed in

```
auto start = parseAttr(getattr(sliceObject, "start"));
auto stop = parseAttr(getattr(sliceObject, "stop"));
auto step = parseAttr(getattr(sliceObject, "step"));
```

and then your implementation will depend heavily on your custom class, but I believe you should return an array based on those indices (being careful about `NoneType` arguments). So if you have an underlying array of data, then you might use a couple of for loops to grab the correct elements based on `parseAttr`.
